### PR TITLE
Chore: Upgrade contracts to Solidity 0.5.8

### LINKED ISSUE #80: Migrate contracts codebase to Solidity 0.5
Please explore which patch version should we use having in mind its adoption and stability

### PR DESCRIPTION
**Pull request by facuspagnuolo**
**MERGE AFTER #108**

Fixes #80 and #23

This PR includes:
- Upgrade to solc 0.5.8
- Organize contract files dropping unused mocks and libs
- Add lint task to the CI

I recommend reviewing by commit

⚠️ I decided to upgrade to 0.5.8 since it has been released 4 months ago, and there are only 3 patches on top of it with minor fixes and optimizations. Moreover, I extended some of the contracts from aragonOS to support being compiled with 0.5.x (see [branch](https://github.com/aragon/aragonOS/tree/alpha_upgrade_solidity_5))

### CODE REVIEW

File: contracts/lib/BytesHelpers.sol [hunk 1]
```diff
@@ -8,7 +8,15 @@ library BytesHelpers {
 
         if (length < 4) {
```
Comment (izqui) [inline contracts/lib/BytesHelpers.sol]: As discussed offline, we can remove some of this logic by just returning bytes4(0) if the length is less than 4 bytes.

Also, we could now use `abi.decode` to decode the data (although it doesn't seem to work when the values are not packed :()

File: contracts/lib/BytesHelpers.sol [hunk 2]
```diff
@@ -8,7 +8,15 @@ library BytesHelpers {
...
             uint256 shiftingPositions = 8 - length * 2;
-            result = bytes4(uint256(result) / (16 ** shiftingPositions));
+            result = toBytes4(toUint256(result) / (16 ** shiftingPositions));
         }
     }
+
+    function toBytes4(uint256 i) private pure returns (bytes4 o) {
+        assembly { o := mload(add(i, 0x20)) }
```
Comment (izqui) [inline contracts/lib/BytesHelpers.sol]: Cannot use `mload` as `i` is already in the stack, `o := i` would do the trick

File: contracts/standards/arbitration/IArbitrable.sol [hunk 3]
```diff
@@ -1,7 +1,7 @@
-pragma solidity ^0.4.15;
+pragma solidity ^0.5.8;
 
 
-interface IArbitrable {
+contract IArbitrable {
```
Comment (izqui) [inline contracts/standards/arbitration/IArbitrable.sol]: What has changed in solidity 0.5 to require this change?

File: contracts/standards/rng/RNG.sol [hunk 4]
```diff
@@ -1,42 +1,44 @@
-/**
-*  @title Random Number Generator Standard
-*  @author Clément Lesaege - <clement@lesaege.com>
-*
-*/
+pragma solidity ^0.5.8;
```
Comment (izqui) [inline contracts/standards/rng/RNG.sol]: We could clean up this file now as it is not used anywhere

File: test/registry/jurors-registry-activatation.js [hunk 5]
```diff
@@ -6,12 +6,11 @@ const JurorsRegistry = artifacts.require('JurorsRegistry')
 const ERC20 = artifacts.require('ERC20Mock')
 const JurorsRegistryOwnerMock = artifacts.require('JurorsRegistryOwnerMock')
 
-const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000'
-
 contract('JurorsRegistry', ([_, juror]) => {
   let registry, registryOwner, ANJ
 
   const MIN_ACTIVE_AMOUNT = bigExp(100, 18)
+  const EMPTY_DATA = '0x0000000000000000000000000000000000000000000000000000000000000000'
```
Comment (izqui) [inline test/registry/jurors-registry-activatation.js]: Why is this change needed? This is no longer empty data but 32 zero bytes

File: contracts/standards/rng/RNG.sol (hunk 4 above)
Comment (facuspagnuolo) [reply contracts/standards/rng/RNG.sol]: Sure, I already dropped this one

File: contracts/standards/arbitration/IArbitrable.sol (hunk 3 above)
Comment (facuspagnuolo) [reply contracts/standards/arbitration/IArbitrable.sol]: The thing is that solc 0.5 now requires all interface's methods to be external

File: contracts/lib/BytesHelpers.sol (hunk 2 above)
Comment (facuspagnuolo) [reply contracts/lib/BytesHelpers.sol]: Sure, this was a brainfart 😂 
As discussed offline, let's drop this case anyway

File: test/registry/jurors-registry-activatation.js (hunk 5 above)
Comment (facuspagnuolo) [reply test/registry/jurors-registry-activatation.js]: Yea, it tests were failing otherwise

Comment (facuspagnuolo) [reply test/registry/jurors-registry-activatation.js]: I already rolled back this, it was because of the issue with the `BytesHelper` lib. I didn't listen to my tests 😅

**Comment by facuspagnuolo** [comment]
@izqui addressed your comments. I think you will prefer reviewing #108 before merging this one cause it is right in the middle. Let me know

File: contracts/arbitration/IArbitrable.sol [hunk 6]
```diff
@@ -1,7 +1,7 @@
-pragma solidity ^0.4.15;
+pragma solidity ^0.5.8;
 
 
-interface IArbitrable {
```
Comment (bingen) [inline contracts/arbitration/IArbitrable.sol]: Why has this been moved out from `standards` folder?

File: contracts/standards/arbitration/IArbitrable.sol (hunk 3 above)
Comment (bingen) [reply contracts/standards/arbitration/IArbitrable.sol]: Can't we then make methods external?

File: contracts/court/Court.sol [hunk 7]
```diff
@@ -837,6 +837,7 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
 
         // Transition the minimum number of terms between the amount requested and the amount actually needed
         uint256 totalFee;
+        CourtConfig storage config = _getConfigSafeAt(termId);
```
Comment (bingen) [inline contracts/court/Court.sol]: Why do we have 2 calls to `_getConfigSafeAt` now?

File: contracts/court/Court.sol [hunk 8]
```diff
@@ -855,7 +856,7 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
             emit NewTerm(termId, msg.sender);
 
             // Add amount of fees to be paid for the transitioned term
-            CourtConfig storage config = _getConfigSafeAt(currentTerm.courtConfigId);
+            config = _getConfigSafeAt(currentTerm.courtConfigId);
```
Comment (bingen) [inline contracts/court/Court.sol]: Shouldn't the parameter here be a term id, instead of a court config id?

File: contracts/court/Court.sol [hunk 9]
```diff
@@ -1263,7 +1264,8 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
         CourtConfig storage config = _getConfigAtDraftTerm(_round);
         // Court terms are assumed to always fit in uint64. Thus, the end term of a round is assumed to fit in uint64 too.
         // For a regular round, its end term is at the end of its appeal confirmation term.
-        return _round.draftTermId + _round.delayedTerms + config.commitTerms + config.revealTerms + config.appealTerms + config.appealConfirmTerms;
+        uint64 draftTerm = _round.draftTermId + _round.delayedTerms;
```
Comment (bingen) [inline contracts/court/Court.sol]: Why are calculating this separately?

File: contracts/court/Court.sol (hunk 8 above)
Comment (bingen) [reply contracts/court/Court.sol]: Besides, these changes are not related to Solidity 5, right? Then it would be good to have them in a separate PR.

File: contracts/court/Court.sol [hunk 10]
```diff
@@ -1543,13 +1553,13 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
     }
 
     // TODO: move to a factory contract
-    function _initSubscriptions(ERC20 _feeToken, uint256[5] _subscriptionParams) private {
-        uint64 MAX_UINT64 = uint64(-1);
-        uint256 MAX_UINT16 = uint16(-1);
+    function _initSubscriptions(ERC20 _feeToken, uint256[5] memory _subscriptionParams) private {
+        uint64 maxUINT64 = uint64(-1);
```
Comment (bingen) [inline contracts/court/Court.sol]: I guess the linter doesn't complain anymore, but this looks awkward to me. Wouldn't it be more common something like `maxUint64`?

File: contracts/lib/BytesHelpers.sol [hunk 11]
```diff
@@ -1,14 +1,14 @@
-pragma solidity ^0.4.24;
+pragma solidity ^0.5.8;
 
 
 library BytesHelpers {
     function toBytes4(bytes memory _self) internal pure returns (bytes4 result) {
         uint256 length = _self.length;
```
Comment (bingen) [inline contracts/lib/BytesHelpers.sol]: This seems unnecessary verbose now.

File: contracts/registry/JurorsRegistry.sol [hunk 12]
```diff
@@ -38,7 +38,7 @@ contract JurorsRegistry is Initializable, IsContract, IJurorsRegistry, ERC900, A
     string internal constant ERROR_TOKEN_APPROVE_NOT_ALLOWED = "JR_TOKEN_APPROVE_NOT_ALLOWED";
 
     // Address that will be used to burn juror tokens
-    address internal constant BURN_ACCOUNT = 0xdead;
+    address internal constant BURN_ACCOUNT = address(0x000000000000000000000000000000000000dEaD);
```
Comment (bingen) [inline contracts/registry/JurorsRegistry.sol]: Interesting. Wasn't Solidity 5 accepting the old form?

File: contracts/subscriptions/CourtSubscriptions.sol [hunk 13]
```diff
@@ -339,8 +339,8 @@ contract CourtSubscriptions is IsContract, ISubscriptions, TimeHelpers {
     }
 
     function _setFeeToken(ERC20 _feeToken) internal {
-        require(isContract(_feeToken), ERROR_NOT_CONTRACT);
-        if (accumulatedGovernorFees > 0) {
+        require(isContract(address(_feeToken)), ERROR_NOT_CONTRACT);
+        if (accumulatedGovernorFees > uint256(0)) {
```
Comment (bingen) [inline contracts/subscriptions/CourtSubscriptions.sol]: See this: https://github.com/aragon/aragon-court/pull/103#discussion_r321798681
Let's make sure we are consistent once we make a decision there.

File: contracts/standards/arbitration/IArbitrable.sol (hunk 3 above)
Comment (facuspagnuolo) [reply contracts/standards/arbitration/IArbitrable.sol]: based on the current `Arbitrable` contract we have

File: contracts/arbitration/IArbitrable.sol (hunk 6 above)
Comment (facuspagnuolo) [reply contracts/arbitration/IArbitrable.sol]: I prefer using the word `standard` to refer interfaces agreed by the whole community (i.e. EIP). We could rename it to `EIP792` as we are doing for the rest of the standards, but I'm not sure if we are sticking it to the point

File: contracts/standards/arbitration/IArbitrable.sol (hunk 3 above)
Comment (bingen) [reply contracts/standards/arbitration/IArbitrable.sol]: But we can have an internal function called from the contract wherever is needed and from the external one, like we do in many other places.

File: contracts/court/Court.sol (hunk 8 above)
Comment (facuspagnuolo) [reply contracts/court/Court.sol]: > Shouldn't the parameter here be a term id, instead of a court config id?

Yes, I think this was carried from [#103](https://github.com/aragon/aragon-court/pull/103/files?file-filters%5B%5D=.sol#diff-4bb317b489994cf6f4c4b7ed11c18590R850). Nice catch!

> These changes are not related to Solidity 5, right? 

In fact, they are. Note that previously we were declaring a storage pointer to the `CourtConfig` inside the for-loop and using it at the end when transferring founds. This is easy to solve declaring the pointer outside the for-loop, but I was getting a warning due to declaring a storage pointer without initializing it. Therefore, I initialized it to point to the current `CourtConfig`.

> Why do we have 2 calls to `_getConfigSafeAt` now?

I think the explanation above answers this quiestion as well

File: contracts/court/Court.sol (hunk 7 above)
Comment (facuspagnuolo) [reply contracts/court/Court.sol]: Answered [here](https://github.com/aragon/aragon-court/pull/109#discussion_r325312162)

File: contracts/court/Court.sol (hunk 9 above)
Comment (facuspagnuolo) [reply contracts/court/Court.sol]: To avoid exceeding the max line length while avoiding to have a multi-line math operation

File: contracts/court/Court.sol (hunk 10 above)
Comment (facuspagnuolo) [reply contracts/court/Court.sol]: No strong feelings, I can change it for now. However, remember this was a change towards #78

File: contracts/registry/JurorsRegistry.sol (hunk 12 above)
Comment (facuspagnuolo) [reply contracts/registry/JurorsRegistry.sol]: Nope, it requires properly formed and checksumed addresses now

File: contracts/subscriptions/CourtSubscriptions.sol (hunk 13 above)
Comment (facuspagnuolo) [reply contracts/subscriptions/CourtSubscriptions.sol]: Fixed in https://github.com/aragon/aragon-court/pull/109/commits/a6c49f0b1c1e23a42ca86c558574d290ec2929b3

File: contracts/lib/BytesHelpers.sol (hunk 11 above)
Comment (facuspagnuolo) [reply contracts/lib/BytesHelpers.sol]: Fixed in https://github.com/aragon/aragon-court/pull/109/commits/a6c49f0b1c1e23a42ca86c558574d290ec2929b3

File: contracts/court/Court.sol (hunk 10 above)
Comment (facuspagnuolo) [reply contracts/court/Court.sol]: Fixed in https://github.com/aragon/aragon-court/pull/109/commits/a6c49f0b1c1e23a42ca86c558574d290ec2929b3

File: contracts/court/Court.sol [hunk 14]
```diff
@@ -223,11 +224,10 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
...
 
     /**
-    * @dev Ensure the msg.sender is a certain address
-    * @param _addr Address expected to be the msg.sender
+    * @dev Ensure the msg.sender is the CR Voting module
     */
-    modifier only(address _addr) {
-        require(msg.sender == _addr, ERROR_INVALID_SENDER_ADDRESS);
+    modifier onlyVoting() {
+        require(msg.sender == address(voting), ERROR_INVALID_SENDER_ADDRESS);
```
[the remark this answers is not in the document]
Comment (facuspagnuolo) [reply contracts/court/Court.sol]: Fixed in https://github.com/aragon/aragon-court/pull/109/commits/a6c49f0b1c1e23a42ca86c558574d290ec2929b3

File: contracts/court/Court.sol (hunk 8 above)
Comment (bingen) [reply contracts/court/Court.sol]: Oh, I see! Yes, too bad that Solidity was allowing that (we have had similar issues in the past, accidentally using a pointer before its declaration, which was allowed :scream: ).
Thanks for the clarification.

File: contracts/court/Court.sol [hunk 15]
```diff
@@ -851,7 +850,7 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
             emit NewTerm(currentTermId, msg.sender);
 
             // Add amount of fees to be paid for the transitioned term
-            config = _getConfigSafeAt(currentTerm.courtConfigId);
+            config = _getConfigSafeAt(currentTermId);
             totalFee = totalFee.add(config.heartbeatFee.mul(uint256(currentTerm.dependingDrafts)));
```
Comment (izqui) [inline contracts/court/Court.sol]: Unrelated to this PR: but this is problematic if the Court config changes in between the terms being transitioned and the fee token is changed. We should fix in a separate issue.

Comment (facuspagnuolo) [reply contracts/court/Court.sol]: Totally, it's weird

Comment (facuspagnuolo) [reply contracts/court/Court.sol]: Created #127
